src/main/scala/org/constellation/consensus/Round.scala

```diff
@@ -0,0 +1,157 @@
+      if (transactionProposals.size >= numberOfTransactionProposalsSent) {
+        Option(unionTransactionProposalsTikTok).foreach(_.cancel())
+        self ! UnionProposals(id)
+      }
+    case ReceivedUnionedBlock(_, facilitatorId, checkpointBlock) =>
+      checkpointBlockProposals += facilitatorId -> checkpointBlock
+      if (checkpointBlockProposals.size >= numberOfCheckpointBlockProposalsSent) {
+        self ! ResolveMajorityCheckpointBlock(id)
+      }
+    case UnionProposals(_)                 => unionProposals()
+    case ResolveMajorityCheckpointBlock(_) => resolveMajorityCheckpointBlock()
+    case msg                               => log.info(s"Received unknown message: $msg")
+  }
+
+  private def scheduleUnionProposals: Cancellable =
+    context.system.scheduler.scheduleOnce(
+    ConfigUtil.getDurationFromConfig("constellation.consensus.union-proposals-timeout", 15.second),
+    self,
+    UnionProposals(id)
+  )
+
+
+  def unionProposals(): Unit = {
+    // TODO: wkoszycki is this the way to merge transactions ?
+    val transactions = transactionProposals.values.flatten.toSet
+    // TODO: wkoszycki should I pull messages and tips this way (services are altering the data itself)?
+    val messages = dao.threadSafeMessageMemPool.pull(1).getOrElse(Seq())
+    val maybeCheckpointBlock = dao
+      .pullTips()
+      .map { tips =>
+        CheckpointBlock.createCheckpointBlock(transactions.toSeq, tips._1.map { soe =>
+          TypedEdgeHash(soe.hash, EdgeHashType.CheckpointHash)
+        }, messages)(dao.keyPair)
+      }
+    maybeCheckpointBlock match {
+      case Some(cb) =>
+        log.info("Created proposals union and broadcasting to others")
+        val readyPeers = dao.readyPeers
+        context.parent ! UnionBlockProposal(id, readyPeers.values, FacilitatorId(dao.id), cb)
+        checkpointBlockProposals += FacilitatorId(dao.id) -> cb
+        numberOfCheckpointBlockProposalsSent = readyPeers.size + 1
+      case None => log.warning("CheckpointBlock was not created during proposals union")
+    }
+  }
+
+  def resolveMajorityCheckpointBlock(): Unit = {
+    implicit val shadedDao: DAO = dao
+    implicit val ec: ExecutionContextExecutor = dao.edgeExecutionContext
+
+    // TODO: wkoszycki what do we do with such condition
+    //      1 -> [A, B, C, D]
+    //      2 -> [E, f, C, D]
+    // TODO: wkoszycki should group by soe or hash ?
+    val sameBlocks = checkpointBlockProposals
+      .groupBy(_._2.soeHash)
+      .maxBy(u => u._2.size)._2
+    log.info(s"Grouped blocks size ${sameBlocks.size}")
+
+    val majorityCheckpointBlock = sameBlocks.values.foldLeft(sameBlocks.head._2)(_ + _ )
+
+    //    broadcast to other peers
+    val cache = CheckpointCacheData(Some(majorityCheckpointBlock), height = majorityCheckpointBlock.calculateHeight())
+    val finalFacilitators = checkpointBlockProposals.keySet.map(f => f.id).toSet
+
+    val responses = dao.peerInfo.values.toList.map { peer =>
```

Comment:
This should only send to the peers that were not facilitators in this round. Otherwise nodes will double accept things. Before we were not accepting blocks on the facilitators until this had been received, but since they now will accept it after consensus this should only broadcast to the nodes that were not chosen as facilitators.